### ISSUE TITLE
(next-urql) Resetting client does not notify query listeners

### ISSUE THREAD

**Issue by AlexanderArvidsson** [FormidableLabs/urql#2281]
After resetting the client using `resetUrqlClient`, I would expect that all existing data in queries would be reset.
This does not seem to be the case, my `useQuery` hooks are still returning the data.

In my use case, I have a user token that is used to fetch the user data from the backend, but only when they are signed in.
I use `pause` to delay the request to only when the user is logged in.
When logging out, I clear the token, and `pause` becomes true.
It does not seem that resetting clients will clear the data of queries that are paused (I do not know if this is the reason, difficult to test).

**urql version & exchanges:** 
urql version: 2.1.3
next-urql version: 3.3.2
exchanges:
1. devtoolsExchange
2. dedupExchange
3. requestPolicyExchange
4. refocusExchange
5. ssrExchange
6. cache Exchange (graphcache) 
7. authExchange
8. fetchExchange

Note that the code sandbox uses the pokedex api instead of the above exchanges. 

**Steps to reproduce**
1. Set up a query that fetches data
2. Call `resetUrqlClient`
3. Observe that any previous successful queries are still keeping their data locally in the `useQuery` hook

**Expected behavior**
Resetting the client should notify all queries to reset their data to undefined.

**Actual behavior**
`useQuery` hook persists their data, even though the client was reset.
One possibility for this could be, as previously mentioned, the fact that my queries are `pause`d after resetting the client.
Devtools does show that the data is cleared from cache correctly

**Codesandbox Reproduction**
1. Click the login button to "simulate" a user login with token
2. Data will be fetched from the API
3. Click the logout button to "simulate" logging out, clearing their token
4. Data is still there, and stale is false, even though the client was reset.
https://codesandbox.io/s/next-js-forked-qzpiht?file=/pages/index.js

**Comment by kitten** [comment]
I looked at the reproduction a little and noticed these lines:

```js
const [res] = useQuery({
  query: queryPokemon,
  pause: !token
});

const onClick = () => {
  resetUrqlClient();
  setToken(undefined);
}
```

Basically, as you're describing:
> I use pause to delay the request to only when the user is logged in.

The problem here is when you're resetting your client you're resetting the state of `token` to be null-ish, which means that while the client is changing, your `useQuery` hook won't be subscribed and won't be able to execute/accept a new query result here. So, that's basically expected behaviour.

The issue here is that authentication state rarely does well together with UI state (the hooks being in UI state territory, and your login / auth state being in global/data territory)

So, you'll have to be careful here to keep them apart by a well margin. For instance, when you're logging a user in, and you keep that auth state somewhere, for instance in a manual provider or together with `authExchange`, then you may want to simply accept that some queries can fail with auth errors (or intentionally return null-ish results in your GraphQL schema)

There are basically many solutions here, but they come down to either:
- changing the schema to account for logged out queries
- ensuring that the UI is more aware of when a login/logout has occured and rendering components based on that
- reloading/transitioning the page entirely when login/logout occurs

But basically, if you assume that your login state won't live in the local component, it's likely that you'll bounce into one of the above solutions "naturally".

That said, it's also worth noting that `pause` here is useful for doing something based on UI state (e.g. form validation) but it's rarely good for business logic or to simulate some kind of global state locally. The reason for that is that `pause` _literally_ is a "blunt tool" that stops anything from changing in `useQuery`. It stops everything but keeps the state intact.

It's not quite the same as, for instance, pausing until a new query comes in. If you were to really want that, you'll probably have to reach for more state that triggers `executeQuery` manually 🤔

**Comment by AlexanderArvidsson** [comment]
@kitten I currently have the authentication state inside authExchange already, my reproduction was a very big simplification.

What I want to avoid is calling the backend when I **know** the user is not authenticated, because that would put unnecessary load on the backend.

I did think about your second point, where UI is not rendered when you are not logged in. That would solve the issue, but I also want the user model to exist within the application context at all times, so I don't have to query for the user every time want it (cache-only could work for this though, and I could trigger refresh at certain points manually).

From what I gather from your answer, there is no way to clear the cached data within a user query except literally unmounting the component and re-mounting. That sounds like an oversight to me?

Other libraries, such as RTK-Query, utilizes a "skip token" which in combination with their hook, returns the data differently.
They have `data` and `currentData` in the hook, where `data` could be stale (from cache) and `currentData` is never stale (it could come from cache, but they have refetch policies that ensure this is never stale).
When passing a skip token, the `currentData` becomes undefined, because the arguments to the query has changed (i.e. the data would be stale relative to the query you made).
In other words, `currentData` _is never stale relative to the query you made_.
I want something similar. If I tell my hook to pause, I know my data is stale, and as such I do not want the data there.

Back to `urql`, what would you suggest I do to have the user model always be available on all pages (I am using a context), but on logout this needs to be cleared? I do not count "refresh the page" as a sufficient solution to this problem.

**Comment by kitten** [comment]
Not really what I meant, no. What I meant was that the problem here is simply triggered by the use of local state overriding a potential reload. We can't change that.

If you're after avoiding requests, the tool to go with is generally the normalised cache where you can selectively invalidate or update data proactively. But from a request standpoint, if you pause a query in the UI, it won't update 😅 That's completely separate from reinstantiating the client or for making or not making requests.

In other words, it's simply not a UI concern whether to actually send an API request or not. All we can do in the UI is say that we want data for a given operations, and that's very intentional.
If you do only ever want stale data, then that's what the `requestPolicy` is for, so alternatively, you could try to use a `cache-only` policy.
https://formidable.com/open-source/urql/docs/basics/document-caching/#request-policies

If I'm not mistaken, if you want to avoid a request from the UI-end in some cases that could be it? 🤔 I may be misunderstanding this though
It's important to note that React Query and urql aren't quite an apple to apples comparison. We simply have very different concepts and architectural choices, especially because we clearly differentiate between the UI and data requirements (which are closely related to the bindings) and data/global state which is related more to the client and the operations 😅

**Comment by AlexanderArvidsson** [comment]
The original point of the issue really did not have much to do with that I want to avoid making API requests, it is that there is currently no way of resetting the UI once you sign out without refreshing the page / remounting the components.

If the user logs out, I need to make sure the UI does not show that you are still logged in, and resetting the client will do exactly that, but only in the "data layer". The UI will not know that it has been reset.

Now, the codesandbox might be misrepresentative, since it always returns data regardless if you are logged in.
In my case, the GraphQL query returns an error when you are unauthenticated.
Making a request to the user endpoint when I know the user is not signed in is redundant, which is why I used pause for to avoid that.

If we take one step back and consider the case where we have a user endpoint that returns the currently signed in user (no pause).
When the user logs out and the client is reset, the client will make _another_ request to this endpoint which will return an error.
In this case, the data would become undefined, but only after the new request is completed. We are still making an unnecessary request when we already **know** that the user is not logged in anymore.

Is your recommendation that I let my hooks call the user endpoint even though the user is not logged in?
In the case where you have a navbar which shows the currently signed in user, this component would be rendered on every page, making requests to the API which ultimately returns errors.
Conditionally rendering subcomponents of the navbar would solve it, sure, but that's assuming I have access to the authState which I do not seem to have, so how would I even conditionally render based on if I am logged in?
I currently have a round-about way of getting my authState using an event emitter that saves token to state.

TL;DR; How do I solve my problem where the user still shows up in the UI even though I logged out? I need to hide the logged in user when they are not logged in anymore, without forcing a refresh.
